src/main/kotlin/no/skatteetaten/aurora/mokey/service/ApplicationDataService.kt

```diff
@@ -0,0 +1,11 @@
+package no.skatteetaten.aurora.mokey.service
+
+import no.skatteetaten.aurora.mokey.model.ApplicationData
+
+interface ApplicationDataService {
```

Comment:
Trenger vi dette interfacet? Skjer det noen gang at vi skal bruke cache/cluster om hverandre. Akkurat nå har vi død kode i cluster koden fordi vi skal tilfredstille denne kontrakten som vi slik jeg ser det ikke trenger. Eller er det meg som mangler noe info her?

Reply:
Jeg må innrømme at for min del er det et poeng at ApplicationDataServiceCacheDecorator og ApplicationDataServiceOpenShift implementerer samme interface for å gjøre det tydelig CacheDecorator gjør akkurat det samme som OpenShift, bare via en cache.

Det som drev meg litt i den retningen var at jeg syntes ikke det føltes riktig at kontrollerne (f.eks. ApplicationDetailsController) skal injisere CachedApplicationDataService direkte (hvorfor skal kontrollerne forholde seg til implementasjonsdetaljer som caching?). ApplicationDetailsController bør heller ikke ha tilgang til å kalle opp refreshCache-metoden, som er en del av public-interfacet til CachedApplicationDataService. Kontrollerne bør kun trenge forholde seg til de metodene som har med uthenting av data å gjøre. Derfor ApplicationDataService-interfacet.

Jeg syntes også det var viktig å skille håndtering av cache og uthenting av data fra OpenShift i to forskjellige klasser - dette er to helt forskjellige ansvar. Jeg syntes også at designet vårt burde ha mulighet til å bypasse caching dersom vi ønsket det, så det føltes egentlig veldig naturlig å implementere caching som en decorator for ApplicationDataServiceOpenShift. Vi kan da støtte det samme interfacet både ved å gå via cache, og ved å gå direkte mot openshift for hver forespørsel.

Vi kan for så vidt godt gjøre det slik at ApplicationDataServiceOpenShift ikke implementerer ApplicationDataService for å unngå å ha kode som ikke blir brukt når man går via cache. Men, jeg mener likevel at vi bør beholde ApplicationDataService-interfacet for å unngå å eksponere for mye detaljer til kontrollerne. Det er mulig det er bittelitt overengineering å la ApplicationDataServiceOpenShift implementere ApplicationDataService, men de øvrige delene av designet føler jeg har sin plass.

Reply:
Har tenkt litt frem og tilbake på denne. Det er mye av det du sier jeg er enig i. Den siste paragrafen din snakker egentlig for seg selv synes jeg. Hvis man fjerner at OpenShift service implementerer interfacet så er jeg helt med på resten av det.

Reply:
Jeg likte litt måten dette var satt opp på jeg, det ble en clean implementasjon av cache. Hvor cachen bare wrapper seg rundt selve implementasjonen. Det er jo på mange måter det spring-cache gjør også, bare at den er mer generisk.

Hvis man endrer slik at ApplicationDataServiceOpenShift ikke implementerer ApplicationDataService lenger, da mister man vel den muligheten til å enkelt kjøre uten cache? Er dette noe vi ønsker å ha mulighet til å gjøre?